dali/pipeline/executor/executor2/op_task.cc

```diff
@@ -0,0 +1,366 @@
+
+  auto *input_edge = node_->inputs[input_idx];
+  auto &source = input_edge->producer->outputs[input_edge->producer_output_idx];
+  if (!source.parallel_consumers) {
+    // If there's just one consumer, we can just set the layout
+    input.SetLayout(layout);
+    if (source.consumers.size() > 1_uz) {
+      // There are multiple (sequential) consumers - we must reset the layout for the next consumer
+      reset_input_layouts_.push_back(input_idx);
+    }
+  } else {
+    // If there are multiple consumers, then we have to create a new object
+    // sharing the data, as setting it in-place might cause a race condition.
+    auto new_input = std::make_shared<TensorList<Backend>>();
+    new_input->ShareData(input);
+    new_input->SetLayout(layout);
+    ws_->SetInput(input_idx, std::move(new_input));
+  }
+}
+
+void OpTask::ResetInputLayouts() {
+  for (int i : reset_input_layouts_) {
+    if (ws_->InputIsType<CPUBackend>(i)) {
+      ws_->UnsafeMutableInput<CPUBackend>(i).SetLayout({});
+    } else {
+      assert(ws_->InputIsType<GPUBackend>(i));
+      ws_->UnsafeMutableInput<GPUBackend>(i).SetLayout({});
+    }
+  }
+}
+
+
+void OpTask::ApplyDefaultLayouts() {
+  auto &schema = node_->op->GetSpec().GetSchema();
+  for (int i = 0; i < ws_->NumInput(); i++) {
+    if (ws_->InputIsType<CPUBackend>(i)) {
+      ApplyDefaultLayout<CPUBackend>(i, schema);
+    } else {
+      assert(ws_->InputIsType<GPUBackend>(i));
+      ApplyDefaultLayout<GPUBackend>(i, schema);
+    }
+  }
+}
+
+void OpTask::SetupOp() {
+  auto &ws = *ws_;
+  int nout = node_->outputs.size();
+  assert(ws.NumOutput() == nout);
+
+  skip_ = ShouldSkip();
+
+  int device = -1;
+
+  for (int i = 0; i < nout; i++) {
+    if (ws.OutputIsType<CPUBackend>(i)) {
+      if (!ws.OutputPtr<CPUBackend>(i)) {
+        auto tl = std::make_shared<TensorList<CPUBackend>>();
+        bool pinned = node_->outputs[i].pinned;
+        tl->set_pinned(pinned);
+        if (pinned) {
+          tl->set_order(ws.output_order());
+          if (device < 0)
+            CUDA_CALL(cudaGetDevice(&device));
+          tl->set_device_id(device);
+        }
+        ws.SetOutput(i, tl);
+      }
+    } else if (ws.OutputIsType<GPUBackend>(i)) {
+      if (!ws.OutputPtr<GPUBackend>(i)) {
+        auto tl = std::make_shared<TensorList<GPUBackend>>();
+        tl->set_order(ws.output_order());
+        if (device < 0)
+          CUDA_CALL(cudaGetDevice(&device));
+        tl->set_device_id(device);
+        ws.SetOutput(i, tl);
+      }
+    } else {
+      assert(!"Unreachable code - unknown backend.");
+    }
+  }
+
+  if (!skip_) {
+    DomainTimeRange tr("[DALI][OpTask] Setup " + GetOpDisplayName(node_->op->GetSpec()));
+    ApplyDefaultLayouts();
+    std::vector<OutputDesc> output_descs;
+    output_descs.reserve(nout);
+    // If Setup returns true, we must resize the outputs;
+    if (node_->op->Setup(output_descs, ws)) {
```

Comment:
What about the CanInferOutputs() and cross-validation with it?

Reply:
Isn't CanInferOutputs redundant? I'd gladly remove it instead of proliferating its use.

Reply:
It's used by the old executor to set contiguity when prepopulating/preallocating. This executor has neither preallocation nor prepopulation.
TBH, I was surprised that the old executor enforced that at run-time - it doesn't serve any purpose when the pipeline is already running. We should remove this check and keep CanInferOutputs only for preallocation.